### DISCUSSION TITLE
Trigger click events remotely for a specific user

### DISCUSSION THREAD

**Discussion by EccentricTom** [zauberzeug/nicegui#1836] [Q&A]
### Question

Hi there.

I am developing an application where users can record microphone data. The idea is that certain users are able to do this by themselves, while others are monitored from the backend. I have wxpython GUI that I am using that will do the backend monitoring, and I need to reflect button presses made in one GUI in the other. I have been able to achieve this from the Nicegui page to the wxpython GUI with a WebSocket connection, but I have not achieved the reverse except for disabling Nicegui elements.

I have been trying to find a way to access and trigger click events in a particular session by using the NiceGUI client, and I have been able to trigger ui.notify() on a per session basis:

```python
async def process_command(data):
    command = data.get('command')
    data = data.get('data')
    user_clients = [c for c in globals.clients.values() if
                    c != globals.index_client]
    if command == 'disable':
        from routes.client_recording import cond_disable
        bindings.disable_all()
        for client in user_clients:
            with client:
                ui.notify('disabled')
        await asyncio.sleep(10)
        bindings.enable_all()
        for client in user_clients:
            with client:
                ui.notify('enabled')
        await cond_disable()
```

Here is also an example of how I trigger click events within the wxpython GUI from NiceGUI:
```python
# This is the function that NiceGUI triggers. The connect_clients variable is a set containing the client that is connected via a websocket that allows data to be sent between the wxPython GUI and the NiceGUI backend
class ClientToServer:
    def __init__(self):
        self.router = APIRouter()
        self.router.add_api_route("/start_stop_record", self.start_stop_rec,
                                  methods=["POST"])
    async def start_stop_rec(self):
        message = {
            "type": "websocket.send",
            "text": json.dumps({
                "type": "Event",
                "command": "start/stop record"
            })
        }
        for client in connected_clients:
            await client.send_json(message)

# This is the endpoint in the wxpython GUI and the events it can trigger
async def connect_to_websocket(frame):
    try:
        async with websockets.connect("ws://localhost:8000/session/ws") as websocket:
            frame.websocket = websocket
            while True:
                message = await websocket.recv()
                try:
                    message_json = json.loads(message)
                    message_content = json.loads(message_json.get("text", None))
                    message_type = message_content.get('type', None)
                    message_command = message_content.get('command', None)
                    message_content = message_content.get('data', None)
                    frame.process_message((message_command, message_content)) # This is a function within the wxpython GUI that uses a series of if-elif statements checking the message command and then triggering the correct function.
                except json.JSONDecodeError as e:
                    print(f"Error decoding JSON: {e}")

    except ConnectionRefusedError:
        frame.Destroy()
        
def remote_start_stop_record(frame):
    if frame.panel.record_button.GetValue() is True:
        frame.panel.record_button.SetValue(False)
    elif frame.panel.record_button.GetValue() is False:
        frame.panel.record_button.SetValue(True)
    wx.CallAfter(frame.panel.start_stop_recording, wx.EVT_BUTTON)      

```

However, I need to be able to access the elements within the page and trigger click events that will only affect a specific user. Some variables I can change by accessing `app.user.storage`, but there are some functions that cannot have data bound to the variables in storage. I also would like to have visual feedback for the user when a button is being clicked. 

Alternatively, I have considered making all inner page functions fastAPI requests, but I am not sure how feasible this is and worried that such an approach would have too much of a delay. This would also not give the visual feedback I would like.

So my question is, is there a way for me to trigger click events within the page of a particular user, or is there a better method of achieving this? Thank you in advance!

**Comment by falkoschindler** [comment]
Hi @EccentricTom, if you really need to trigger something from another app, an dedicated FastAPI endpoint sounds reasonable. I don't think there would be a huge delay, but it depends on your network (internet?) connection. Like with any other Python app you could probably also open another socket connection to reduce the latency.

Triggering the visual feedback sounds more challenging. I would use Google/StackOverflow/ChatGPT to find a way to trigger mouse clicks on an arbitrary HTML page so that it feels like a real click for the user, including ripple effects on Quasar elements. If there's a way to do that, it should be possible to do it from the NiceGUI backend.

**Comment by rodja** [comment]
If you do not care about the actual mouse click you could use `app.storage.browser['id']` to identify a user and simply call model changes for this user.

**Reply by rodja** [reply]
Something remotely resembling might be our [chat app example](https://github.com/zauberzeug/nicegui/blob/main/examples/chat_app/main.py) where we have a `messages` model which is shared by all users.

**Comment by EccentricTom** [comment]
Thank you @falkoschindler and @rodja for taking the time to answer my question. I know that my use case is very niche, and possibly counterintuitive.

What I have found that could potentially work to at least trigger the functions within a page is by calling the elements within the client stored within nicegui.globals. For example:

```python
  user_clients = [c for c in globals.clients.values() if
                    c != globals.index_client]
for client in user_clients:
    element = client.elements.get(5, None)
    element_listener = element._event_listeners.get(listener_id, None)
    function = element_listener.handler
    function(element)
```

Of course my final solution would be more refined than this, and might be overly complicated as compared to using an endpoint, but I found it interesting.

As for triggering a mouse event, I think that it is possible by passing the following javascript into ui.run_javascript:
```js
// Find the element you want to click
var element = document.querySelector("your-element-selector");

// Simulate a click event
var event = new MouseEvent('click', {
    bubbles: true,
    cancelable: true,
    view: window
});
element.dispatchEvent(event);
```
Which should trigger any visual effects of a Quasar element but I have yet to test it
